### PR TITLE
Use hubsender from storage like all other market participants

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

This is yet another step towards using the actor data from the (shared) actor register.

The hub sender configuration has been removed and replaced with a seeded hub sender in the database.
All usages of the configuration have been updated accordingly. The work, however, is more massive than one might have anticipated. This is due to the desire to avoid the "local message hub" having to load data from the command model. This way we keep a low coupling between the two parts of the system.

## References

* #924

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Model/202201121500 Store sender of available data.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    ADD [SenderRole] INT NOT NULL DEFAULT(7);
+GO
+
+ALTER TABLE [MessageHub].[AvailableChargeReceiptData]
+    ADD [SenderId] NVARCHAR(35) NOT NULL DEFAULT('5790001330552');
+GO
+
+ALTER TABLE [MessageHub].[AvailableChargeReceiptData]
+    ADD [SenderRole] INT NOT NULL DEFAULT(7);
+GO
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Model/202201121500 Store sender of available data.sql:31]: Is it correct to use DEFAULT here? Isnt it more correct just to update the current rows with these values and then make it NOT NULL afterwards?

Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Model/202201121500 Store sender of available data.sql:31]: Sure. Thank you. I'll fix it.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeData/AvailableChargeDataFactoryTests.cs [hunk 2]
```diff
@@ -48,17 +49,18 @@ public class AvailableChargeDataFactoryTests
...
+            marketParticipantRepository
+                .Setup(r => r.GetHubSenderAsync())
+                .ReturnsAsync(hubSenderBuilder.Build());
+
+            messageMetaDataContext.Setup(m => m.RequestDataTime).Returns(now);
 
             // Act
-            var actualList =
-                await sut.CreateAsync(acceptedEvent);
+            var actualList = await sut.CreateAsync(acceptedEvent);
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeData/AvailableChargeDataFactoryTests.cs:63]:
```suggestion
            var actual = await sut.CreateAsync(acceptedEvent);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeLinkReceiptData/AvailableChargeLinkReceiptDataFactoryTests.cs [hunk 3]
```diff
@@ -37,23 +38,21 @@ public class AvailableChargeLinkReceiptDataFactoryTests
...
-                .Returns(now);
-
+            messageMetaDataContext.Setup(m => m.RequestDataTime).Returns(now);
             var expectedLinks = acceptedEvent.ChargeLinksCommand.ChargeLinks.ToList();
+            marketParticipantRepository.Setup(r => r.GetHubSenderAsync()).ReturnsAsync(hubSenderBuilder.Build());
 
             // Act
-            var actualList =
-                await sut.CreateAsync(acceptedEvent);
+            var actualList = await sut.CreateAsync(acceptedEvent);
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeLinkReceiptData/AvailableChargeLinkReceiptDataFactoryTests.cs:55]:
```suggestion
            var actual = await sut.CreateAsync(acceptedEvent);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeLinksData/AvailableChargeLinksDataFactoryTests.cs [hunk 4]
```diff
@@ -47,23 +49,22 @@ public class AvailableChargeLinksDataFactoryTests
...
-                    m => m.RequestDataTime)
-                .Returns(now);
+            messageMetaDataContext.Setup(m => m.RequestDataTime).Returns(now);
 
             var expectedLinks = acceptedEvent.ChargeLinksCommand.ChargeLinks.ToList();
 
             // Act
-            var actualList =
-                await sut.CreateAsync(acceptedEvent);
+            var actualList = await sut.CreateAsync(acceptedEvent);
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeLinksData/AvailableChargeLinksDataFactoryTests.cs:67]:
```suggestion
            var actual = await sut.CreateAsync(acceptedEvent);
```

**Review by lasrinnil — APPROVED** [review]
LGTM! Left a few comments, reach out if you want to discuss them :)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Model/202201121500 Store sender of available data.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
+ALTER TABLE [MessageHub].[AvailableChargeData]
+    ADD [SenderId] NVARCHAR(35) NOT NULL DEFAULT('5790001330552');
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Model/202201121500 Store sender of available data.sql:2]: Ok. I use `@data_hub`.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/HubSenderMarketParticipant/HubSenderMarketParticipant.cs [hunk 6]
```diff
@@ -13,28 +13,21 @@
...
-namespace GreenEnergyHub.Charges.Infrastructure.Configuration
+namespace GreenEnergyHub.Charges.Domain.HubSenderMarketParticipant
 {
-    public class HubSenderConfiguration : IHubSenderConfiguration
+    public class HubSenderMarketParticipant : MarketParticipant
     {
-        private readonly MarketParticipant _marketParticipant;
-
-        public HubSenderConfiguration(string senderId, MarketParticipantRole senderRole)
+        public HubSenderMarketParticipant(Guid id, string marketParticipantId, bool isActive, IEnumerable<MarketParticipantRole> roles)
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/HubSenderMarketParticipant/HubSenderMarketParticipant.cs:24]: That's not determined by the ctor input.
It's ensured by line 32 of MarketParticipant.cs.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeDataFactory.cs [hunk 7]
```diff
@@ -52,8 +53,11 @@ public async Task<IReadOnlyList<AvailableChargeData>> CreateAsync(ChargeCommandA
                         operation.Points
                             .Select(x => new AvailableChargeDataPoint(x.Position, x.Price))
                             .ToList();
+                    var sender = await GetSenderAsync().ConfigureAwait(false);
 
                     result.Add(new AvailableChargeData(
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeDataFactory.cs:58]: Perhaps. Do you want me to address that in another PR?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableData/AvailableDataBase.cs [hunk 8]
```diff
@@ -25,25 +25,29 @@ namespace GreenEnergyHub.Charges.MessageHub.Models.AvailableData
...
+            string senderId,
+            MarketParticipantRole senderRole,
             string recipientId,
             MarketParticipantRole recipientRole,
             BusinessReasonCode businessReasonCode,
             Instant requestDateTime,
             Guid availableDataReferenceId)
         {
             Id = Guid.NewGuid();
+            SenderId = senderId;
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableData/AvailableDataBase.cs:37]: I don't feel any urge at the moment though I agree that there are quite many args.
I'll do it if you want it.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/HubSenderMarketParticipants/HubSenderMarketParticipantTests.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    public class HubSenderMarketParticipantTests
+    {
+        [Theory]
+        [InlineAutoMoqData]
+        public void Ctor_WhenNoRoles_ThrowsArgumentException(
+            Guid anyGuid,
+            string anyMarketParticipantId)
+        {
+            // ReSharper disable once CollectionNeverUpdated.Local
+            var emptyRoles = new List<MarketParticipantRole>();
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/HubSenderMarketParticipants/HubSenderMarketParticipantTests.cs:33]: But then I can't state my intent by the name `emptyRoles`. But I'll do it if you prefer it.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeLinkReceiptData/AvailableChargeLinkReceiptDataFactoryTests.cs (hunk 3 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeLinkReceiptData/AvailableChargeLinkReceiptDataFactoryTests.cs:55]: Fixed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/HubSenderMarketParticipants/HubSenderMarketParticipantTests.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/HubSenderMarketParticipants/HubSenderMarketParticipantTests.cs:33]: You could inline the list and get around not having to have the comment, and then intentionally empty the collection - thereby showing the intent.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableData/AvailableDataBase.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableData/AvailableDataBase.cs:37]: I'm not sure but if we always use the two together, then I believe it makes sense.

[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableData/AvailableDataBase.cs:37]: Maybe split it into OperationData and then whatever makes sense for the "other" parameters could be a way to make sense of it all.

image

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/HubSenderMarketParticipants/HubSenderMarketParticipantTests.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/HubSenderMarketParticipants/HubSenderMarketParticipantTests.cs]: Fixed

**Pull request merged**